### PR TITLE
Implement digest based md5 logic instead of hexdigest

### PR DESCRIPTION
**Pull request by RomainBoulay**
I noticed some file hashing hiccups with the remote caching logic that I'm addressing in this pull request. In more details, this PR:

- Improves `verify_upload` url and resource usage
- Updates CacheRemoteStorage to add the verify upload flow
- Implements digest based md5 logic instead of hexdigest (previously implemented)
- Refactors CacheRemoteStorage to simplify testing
- Adds unit tests to cover the updated logic

### CODE REVIEW

File: Sources/TuistCache/Cache/URL+Cache.swift [hunk 1]
```diff
@@ -3,27 +3,41 @@ import Foundation
 extension URL {
     static func apiCacheURL(hash: String,
```
Comment (pepibumur) [inline Sources/TuistCache/Cache/URL+Cache.swift:4]: Considering these methods are only used by the factory, I think having a URL extension adding methods to URL that is scoped to one file is not necessary. I'd rather move these methods to the file where they are scoped to.

File: Sources/TuistCache/Cache/URL+Cache.swift [hunk 2]
```diff
@@ -3,27 +3,41 @@ import Foundation
 extension URL {
```
Comment (pepibumur) [inline Sources/TuistCache/Cache/URL+Cache.swift:3]: I think I've mentioned it in the past, but I don't think the responsibility of the methods below should include throwing if the URL is valid. That's something I think should be validated earlier in the process, and at this point, you should assume that the URL is right and therefore remove all the throws.

File: Sources/TuistCache/Cache/Resources/CloudCacheResourceFactory.swift [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        let url = try URL.apiCacheVerifyUploadURL(
+            hash: hash,
+            cacheURL: cloudConfig.url,
+            projectId: cloudConfig.projectId,
+            contentMD5: contentMD5
+        )
+        return jsonResource(for: url, httpMethod: "POST")
+    }
+
+    private func jsonResource<R: Decodable>(for url: URL, httpMethod: String) -> HTTPResource<CloudResponse<R>, CloudResponseError> {
```
Comment (pepibumur) [inline Sources/TuistCache/Cache/Resources/CloudCacheResourceFactory.swift:58]: This is a good candidate to be part of `HTTPResource.swift`.

**Review by pepibumur — COMMENTED** [review]
Good job @RomainBoulay 👏

File: Package.resolved [hunk 4]
```diff
@@ -51,8 +51,8 @@
         "repositoryURL": "https://github.com/krzyzanowskim/CryptoSwift.git",
         "state": {
           "branch": null,
-          "revision": "e2bc81be54d71d566a52ca17c3983d141c30aa70",
```
[the remark this answers is not in the document]
Comment (RomainBoulay) [reply Package.resolved:54]: I could remove that change from this PR but, just to explain, I could not build `tuist` on my machine without this change. SPM was acting weird, the `Stencil` package was not missing... I reset the SPM cache, re-fetched the packages, etc. Nothing worked. Only updating this package did the trick. I am not sure why :/

File: Sources/TuistCache/Cache/CacheRemoteStorage.swift [hunk 5]
```diff
@@ -26,37 +26,37 @@ enum CacheRemoteStorageError: FatalError, Equatable {
 public final class CacheRemoteStorage: CacheStoring {
     // MARK: - Attributes
 
-    private let cloudConfig: Cloud
     private let cloudClient: CloudClienting
     private let fileClient: FileClienting
     private let fileArchiverFactory: FileArchivingFactorying
+    private let cloudCacheResponseFactory: CloudCacheResourceManufacturing
```
[the remark this answers is not in the document]
Comment (RomainBoulay) [reply Sources/TuistCache/Cache/CacheRemoteStorage.swift:32]: `CloudCacheResourceFactory` is the name of the concrete type while `CloudCacheResourceManufacturing` is the name of the `protocol`. The reason is that `Factorying` is not an English term, but `Manufacturing` is.

That said I don't have opinions, if you prefer to rename it, let me know which term you prefer!

File: Sources/TuistCache/Cache/Resources/CloudCacheResource.swift [hunk 6]
```diff
@@ -0,0 +1,8 @@
+import TuistCore
+import TuistSupport
+
+typealias CloudExistsResource = HTTPResource<CloudResponse<CloudHEADResponse>, CloudHEADResponseError>
```
[the remark this answers is not in the document]
Comment (RomainBoulay) [reply Sources/TuistCache/Cache/Resources/CloudCacheResource.swift:4]: Sure

File: Sources/TuistCache/Cache/URL+Cache.swift (hunk 1 above)
Comment (RomainBoulay) [reply Sources/TuistCache/Cache/URL+Cache.swift:4]: Ok, I'll move them

File: Sources/TuistCache/Cache/Resources/CloudCacheResourceFactory.swift (hunk 3 above)
Comment (RomainBoulay) [reply Sources/TuistCache/Cache/Resources/CloudCacheResourceFactory.swift:58]: Good point!

File: Sources/TuistCache/Cache/URL+Cache.swift (hunk 2 above)
Comment (RomainBoulay) [reply Sources/TuistCache/Cache/URL+Cache.swift:3]: Ok, I removed that assertion and the associated logic

File: Sources/TuistCache/Cache/Resources/CloudCacheResourceFactory.swift [hunk 7]
```diff
@@ -0,0 +1,63 @@
+import Foundation
+import TuistCore
+import TuistGraph
+import TuistSupport
+
+protocol CloudCacheResourceManufacturing {
```
[the remark this answers is not in the document]
Comment (RomainBoulay) [reply Sources/TuistCache/Cache/Resources/CloudCacheResourceFactory.swift:13]: Done

File: Package.resolved (hunk 4 above)
[the remark this answers is not in the document]
Comment (pepibumur) [reply Package.resolved:54]: That's odd. If CI passes it's ok.

File: Sources/TuistCache/Cache/CacheRemoteStorage.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (pepibumur) [reply Sources/TuistCache/Cache/CacheRemoteStorage.swift:32]: I'd prefer to rename it to be consistent.

**Review by pepibumur — APPROVED** [review]

File: Sources/TuistCache/Cache/CacheRemoteStorage.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (RomainBoulay) [reply Sources/TuistCache/Cache/CacheRemoteStorage.swift]: Ok, I renamed it to `CloudCacheResourceFactorying`. I'd just recommend renaming all `*Factorying` to `*Manufacturing` at some point in the future!

**Pull request merged**